File: src/reduct/storage/query/query.cc
// Copyright 2023 ReductStore
// This Source Code Form is subject to the terms of the Mozilla Public
//    License, v. 2.0. If a copy of the MPL was not distributed with this
//    file, You can obtain one at https://mozilla.org/MPL/2.0/.

#include "reduct/storage/query/query.h"

#include <google/protobuf/timestamp.pb.h>
#include <google/protobuf/util/time_util.h>

#include <vector>

#include "reduct/config.h"
#include "reduct/storage/block_manager.h"
#include "reduct/storage/io/async_reader.h"

namespace reduct::storage::query {

using core::Error;
using core::Result;
using core::Time;

using io::AsyncReaderParameters;

using google::protobuf::Timestamp;
using google::protobuf::util::TimeUtil;

struct QueryInfo {
  Time start;
  Time stop;
  std::optional<Time> next_record;
  Time last_update;

  query::IQuery::Options options;
};

static google::protobuf::Timestamp FromTimePoint(const Time& time) {
  auto microseconds = std::chrono::duration_cast<std::chrono::microseconds>(time.time_since_epoch()).count();
  return TimeUtil::MicrosecondsToTimestamp(microseconds);
}

static Time ToTimePoint(const google::protobuf::Timestamp& time) {
  return Time() + std::chrono::microseconds(TimeUtil::TimestampToMicroseconds(time));
}

std::vector<int> QueryRecordsInBlock(const IBlockManager::BlockSPtr& block, const QueryInfo& query_info,
                                     const Timestamp& start_ts, const Timestamp& stop_ts) {
  std::vector<int> records;
  records.reserve(block->records_size());
  for (auto record_index = 0; record_index < block->records_size(); ++record_index) {
    const auto& record = block->records(record_index);
    bool in_time_interval =
        record.timestamp() >= start_ts && record.timestamp() < stop_ts && record.state() == proto::Record::kFinished;
    if (!in_time_interval) continue;

    // check if a record has value of labels that match include
    if (!query_info.options.include.empty()) {
      auto include = query_info.options.include;
      for (const auto& label : record.labels()) {
        if (include.contains(label.name()) && label.value() == include[label.name()]) {
          include.erase(label.name());
        }
      }

      if (!include.empty()) continue;
    }

    // check if a record has value of labels that does not match exclude
    if (!query_info.options.exclude.empty()) {
      auto exclude = query_info.options.exclude;
      for (const auto& label : record.labels()) {
        if (exclude.contains(label.name()) && label.value() == exclude[label.name()]) {
          exclude.erase(label.name());
        }
      }

      if (exclude.empty()) continue;
    }

    records.push_back(record_index);
  }

  return records;
}

class HistoricalQuery : public IQuery {
 public:
  HistoricalQuery(const Time& start, const Time& stop, const Options& options) {
    info_ = QueryInfo{
        .start = start,
        .stop = stop,
        .next_record = std::nullopt,
        .last_update = Time::clock::now(),
        .options = options,
    };
  }

  Result<NextRecord> Next(const std::set<google::protobuf::Timestamp>& blocks, IBlockManager* block_manager) override {
    info_.last_update = Time::clock::now();

    auto start_ts = FromTimePoint(info_.start);
    if (info_.next_record) {
      start_ts = FromTimePoint(*info_.next_record);
    }
    auto stop_ts = FromTimePoint(info_.stop);

    // Find start block
    auto start_block_it = blocks.upper_bound(start_ts);
    if (start_block_it == blocks.end()) {
      start_block_it = std::prev(start_block_it);
    } else if (start_block_it != blocks.begin()) {
      start_block_it = std::prev(start_block_it);
    }

    std::vector<int> records;
    IBlockManager::BlockSPtr block;
    Error search_err = Error::kOk;
    for (auto it = start_block_it; it != std::end(blocks) && *it < stop_ts; it++) {
      auto ret = block_manager->LoadBlock(*it);
      if (ret.error) {
        search_err = std::move(ret.error);
        break;
      }

      block = ret.result;
      if (block->invalid()) {
        continue;
      }
      // Find records
      records = QueryRecordsInBlock(block, info_, start_ts, stop_ts);
      if (!records.empty() || info_.options.include.empty()) {
        break;
      }
    }

    RETURN_ERROR(search_err);

    if (records.empty()) {
      return Error::NoContent();
    }

    auto get_timestamp = [block](int index) { return block->records(index).timestamp(); };
    std::ranges::sort(records, {}, get_timestamp);
    auto& record_index = records[0];

    bool last = false;
    if (records.size() > 1) {
      info_.next_record = ToTimePoint(get_timestamp(records[1]));
    } else {
      // Only one record in current block check next one
      auto next_block_it = std::next(blocks.find(block->begin_time()));
      if (next_block_it != std::end(blocks)) {
        if (*next_block_it < stop_ts) {
          info_.next_record = ToTimePoint(*next_block_it);
        } else {
          // no records in next block
          last = true;
        }
      } else {
        // no next block
        last = true;
      }
    }

    auto [reader, reader_err] =
        block_manager->BeginRead(block, AsyncReaderParameters{.path = BlockPath(block_manager->parent_path(), *block),
                                                              .record_index = record_index,
                                                              .chunk_size = kDefaultMaxReadChunk,
                                                              .time = ToTimePoint(get_timestamp(record_index))});
    RETURN_ERROR(reader_err)

    NextRecord next_record{
        .reader = reader,
        .last = last,
    };

    return next_record;
  }

  [[nodiscard]] bool is_outdated() const override {
    const auto current_time = Time::clock::now();
    return current_time - info_.last_update > info_.options.ttl;
  }

 private:
  QueryInfo info_;
};

Result<IQuery::UPtr> IQuery::Build(const std::optional<core::Time>& start, const std::optional<core::Time>& stop,
                                   const IQuery::Options& options) {
  return {
      std::make_unique<HistoricalQuery>(start ? *start : Time::min(), stop ? *stop : Time::max(), options),
      Error::kOk,
  };
}
}  // namespace reduct::storage::query

File: reduct_rs/src/query.rs
// Copyright 2023 ReductStore
// This Source Code Form is subject to the terms of the Mozilla Public
//    License, v. 2.0. If a copy of the MPL was not distributed with this
//    file, You can obtain one at https://mozilla.org/MPL/2.0/.

use crate::record::from_system_time;
use crate::Labels;
use chrono::Duration;
use futures::Stream;
use reduct_base::error::HttpError;
use std::pin::Pin;
use std::task::{Context, Poll};
use std::time::SystemTime;

pub struct Query {
    id: u64,
    start: Option<u64>,
    stop: Option<u64>,
    include: Option<Labels>,
    exclude: Option<Labels>,
    ttl: Option<Duration>,
    continuous: bool,
}

/// Builder for a query request.
pub struct QueryBuilder {
    query: Query,
}

impl QueryBuilder {
    pub(crate) fn new() -> Self {
        Self {
            query: Query {
                id: 0,
                start: None,
                stop: None,
                include: None,
                exclude: None,
                ttl: None,
                continuous: false,
            },
        }
    }

    /// Set the start time of the query.
    pub fn start(mut self, time: SystemTime) -> Self {
        self.query.start = Some(from_system_time(time));
        self
    }

    /// Set the start time of the query as a unix timestamp in microseconds.
    pub fn start_us(mut self, time_us: u64) -> Self {
        self.query.start = Some(time_us);
        self
    }

    /// Set the end time of the query.
    pub fn stop(mut self, time: SystemTime) -> Self {
        self.query.stop = Some(from_system_time(time));
        self
    }

    /// Set the end time of the query as a unix timestamp in microseconds.
    pub fn stop_us(mut self, time_us: u64) -> Self {
        self.query.stop = Some(time_us);
        self
    }

    /// Set the labels to include in the query.
    pub fn include(mut self, labels: Labels) -> Self {
        self.query.include = Some(labels);
        self
    }

    /// Set the labels to exclude from the query.
    pub fn exclude(mut self, labels: Labels) -> Self {
        self.query.exclude = Some(labels);
        self
    }

    /// Set TTL for the query.
    pub fn ttl(mut self, ttl: Duration) -> Self {
        self.query.ttl = Some(ttl);
        self
    }

    /// Set the query to be continuous.
    pub fn continuous(mut self) -> Self {
        self.query.continuous = true;
        self
    }

    /// Set the query to be continuous.
    async fn query(mut self) -> Query {
        todo!()
    }
}

impl Stream for Query {
    type Item = Result<crate::Record, HttpError>;

    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
        todo!()
    }

    fn size_hint(&self) -> (usize, Option<usize>) {
        todo!()
    }
}
